src/mlpack/core/data/string_encoding_policies/bag_of_words_encoding_policy.hpp

```diff
@@ -0,0 +1,163 @@
+ * of tokens. If an item of the dataset has the i-th token, then the i-th
+ * coordinate of the corresponding vector is equal to 1, otherwise it's equal to
+ * zero. The order in which the tokens are labeled is defined by the dictionary
+ * used by the StringEncoding class. The encoder writes data either in the
+ * column-major order or in the row-major order depending on the output data
+ * type.
+ */
+class BagOfWordsEncodingPolicy
+{
+ public:
+  /**
+   * The function initializes the output matrix. The encoder writes data
+   * in the column-major order.
+   *
+   * @tparam MatType The output matrix type.
+   *
+   * @param output Output matrix to store the encoded results (sp_mat or mat).
+   * @param datasetSize The number of strings in the input dataset.
+   * @param maxNumTokens The maximum number of tokens in the strings of the 
+                         input dataset (not used).
+   * @param dictionarySize The size of the dictionary.
+   */
+  template<typename MatType>
+  static void InitMatrix(MatType& output,
+                         const size_t datasetSize,
+                         const size_t /* maxNumTokens */,
+                         const size_t dictionarySize)
+  {
+    output.zeros(dictionarySize, datasetSize);
+  }
+
+  /**
+   * The function initializes the output matrix. The encoder writes data
+   * in the row-major order.
+   *
+   * Overloaded function to save the result in vector<vector<ElemType>>.
+   * 
+   * @tparam ElemType Type of the output values.
+   *
+   * @param output Output matrix to store the encoded results.
+   * @param datasetSize The number of strings in the input dataset.
+   * @param maxNumTokens The maximum number of tokens in the strings of the 
+                         input dataset (not used).
+   * @param dictionarySize The size of the dictionary.
+   */
+  template<typename ElemType>
+  static void InitMatrix(std::vector<std::vector<ElemType>>& output,
+                         const size_t datasetSize,
+                         const size_t /* maxNumTokens */,
+                         const size_t dictionarySize)
+  {
+    output.resize(datasetSize, std::vector<ElemType>(dictionarySize));
+  }
+
+  /**
+   * The function performs the bag of words encoding algorithm i.e. it writes
+   * the encoded token to the output. The encoder writes data in the
+   * column-major order.
+   *
+   * @tparam MatType The output matrix type.
+   *
+   * @param output Output matrix to store the encoded results (sp_mat or mat).
+   * @param value The encoded token.
+   * @param line The line number at which the encoding is performed.
+   * @param index The token index in the line.
+   */
+  template<typename MatType>
+  static void Encode(MatType& output,
+                     const size_t value,
+                     const size_t line,
+                     const size_t /* index */)
+  {
+    // The labels are assigned sequentially starting from one.
+    output(value - 1, line) = 1;
```

Comment:
I've often seen bag of words done as a count of the number of each word in a document (assuming the token is a word), but the implementation here just does `1`.  Is there a reason to prefer either?  I don't have a strong opinion, just wondering. :)

Reply:
Yeah, you're right. I'll go with your variant. It provides more information and I think it's always easy to reduce the output to the binary mode.

Actually, we could introduce a parameter into the policy like we did with the tf-idf policy. However, probably it's better to leave this policy as simple as it is.

Reply:
Agreed, simple is fine for now.  We can always add the support later if we want. :+1: